public/index.php

```diff
@@ -0,0 +1,258 @@
+                $commandString,
+                implode(PHP_EOL, $output)
+            ));
+        }
+
+        return $output;
+    };
+
+    $cleanBuildDir = static function () use ($buildDir, $execute) : void {
+        $execute('rm -rf ' . escapeshellarg($buildDir));
+        $execute('mkdir ' . escapeshellarg($buildDir));
+    };
+
+    $cloneRepository = static function (UriInterface $repositoryUri, string $targetPath) use ($execute) : void {
+        $execute(
+            'git clone '
+            . escapeshellarg($repositoryUri->__toString())
+            . ' ' . escapeshellarg($targetPath)
+        );
+    };
+
+    $getBranches = static function (string $repositoryDirectory) use (
+        $runInPath,
+        $execute
+    ) : MergeTargetCandidateBranches {
+        return $runInPath(static function () use ($execute) {
+            $execute('git fetch');
+
+            return MergeTargetCandidateBranches::fromAllBranches(...array_map(function (string $branch) : BranchName {
+                return BranchName::fromName(\Safe\preg_replace(
+                    '/^(?:remotes\\/)?origin\\//',
+                    '',
+                    trim($branch, "* \t\n\r\0\x0B")
+                ));
+            }, $execute('git branch -r')));
+        }, $repositoryDirectory);
+    };
+
+    $createTag = static function (
+        string $repositoryDirectory, BranchName $sourceBranch, string $tagName, string $changelog, SecretKeyId $keyId
+    ) use (
+        $runInPath,
+        $execute
+    ) : void {
+        $tagFileName = \Safe\tempnam(sys_get_temp_dir(), 'created_tag');
+
+        file_put_contents($tagFileName, $changelog);
+
+        $runInPath(static function () use ($sourceBranch, $tagName, $keyId, $tagFileName, $execute) {
+            $execute(sprintf('git checkout "%s"', $sourceBranch->name()));
+            $execute(sprintf(
+                'git tag %s -F %s --cleanup=verbatim --local-user=%s',
+                escapeshellarg($tagName),
+                escapeshellarg($tagFileName),
+                escapeshellarg($keyId->id())
+            ));
+        }, $repositoryDirectory);
+    };
+
+    $push = static function (string $repositoryDirectory, string $symbol, ?string $alias = null) use (
+        $runInPath,
+        $execute
+    ) : void {
+        $runInPath(static function () use ($symbol, $alias, $execute) {
+            $execute(sprintf(
+                'git push origin %s',
+                $alias !== null ? escapeshellarg($symbol) . ':' . escapeshellarg($alias) : escapeshellarg($symbol)
+            ));
+        }, $repositoryDirectory);
+    };
+
+    $importGpgKey = static function (string $keyContents) use ($execute) : SecretKeyId {
+        $keyFileName = \Safe\tempnam(sys_get_temp_dir(), 'imported-key');
+
+        file_put_contents($keyFileName, $keyContents);
+
+        $output = $execute(sprintf('gpg --import %s 2>&1 | grep "secret key imported"', escapeshellarg($keyFileName)));
```

Comment:
I'd run it each time, because the environment is to be re-provisioned anyway.

Importing a key is an idempotent operation, so it won't be problematic.